src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/Tracker.cs

```diff
@@ -3,83 +3,136 @@
+        private readonly Dictionary<string, Entry> _entryByLibraryName = new(StringComparer.OrdinalIgnoreCase);
 
         public void Track(GraphItem<TItem> item)
         {
-            var entry = GetEntry(item);
-            if (!entry.List.Contains(item))
-            {
-                entry.List.Add(item);
-            }
+            GetOrAddEntry(item).AddItem(item);
         }
 
         public bool IsDisputed(GraphItem<TItem> item)
         {
-            return GetEntry(item).List.Count > 1;
+            return TryGetEntry(item)?.HasMultipleItems ?? false;
         }
 
         public bool IsAmbiguous(GraphItem<TItem> item)
         {
-            return GetEntry(item).Ambiguous;
+            return TryGetEntry(item)?.Ambiguous ?? false;
         }
 
         public void MarkAmbiguous(GraphItem<TItem> item)
         {
-            GetEntry(item).Ambiguous = true;
+            GetOrAddEntry(item).Ambiguous = true;
         }
 
         public bool IsBestVersion(GraphItem<TItem> item)
         {
-            var entry = GetEntry(item);
-
-            var version = item.Key.Version;
+            var entry = TryGetEntry(item);
 
-            foreach (var known in entry.List)
+            if (entry is not null)
             {
-                if (version < known.Key.Version)
+                var version = item.Key.Version;
+
+                foreach (var known in entry.Items)
                 {
-                    return false;
+                    if (version < known.Key.Version)
+                    {
+                        return false;
+                    }
                 }
             }
 
             return true;
         }
 
-        public IEnumerable<GraphItem<TItem>> GetDisputes(GraphItem<TItem> item) => GetEntry(item).List;
+        public IEnumerable<GraphItem<TItem>> GetDisputes(GraphItem<TItem> item)
+        {
+            return TryGetEntry(item)?.Items ?? Enumerable.Empty<GraphItem<TItem>>();
+        }
 
         internal void Clear()
         {
-            _entries.Clear();
+            _entryByLibraryName.Clear();
         }
 
-        private Entry GetEntry(GraphItem<TItem> item)
+        private Entry? TryGetEntry(GraphItem<TItem> item)
         {
-            Entry itemList;
-            if (!_entries.TryGetValue(item.Key.Name, out itemList))
+            _entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry);
+            return entry;
+        }
+
+        private Entry GetOrAddEntry(GraphItem<TItem> item)
+        {
+            if (!_entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry))
             {
-                itemList = new Entry();
-                _entries[item.Key.Name] = itemList;
+                entry = new Entry();
+                _entryByLibraryName[item.Key.Name] = entry;
             }
-            return itemList;
+
+            return entry;
         }
 
-        private class Entry
+        private sealed class Entry
```

Comment:
Out of curiosity, what does adding `sealed` on a private class help with?

I know in Java all methods are virtual, so `sealed` allows the compiler to call methods directly, rather than looking them up in a v-table, which has a very marginal perf benefit. But in .NET, virtual methods must be marked explicitly. And given this is a nested private class, nothing else can inherit and extend the class.

I'm curious if regularly adding `sealed` to new code I write is worthwhile.

Reply:
I generally seal things by default, more for documentation than performance. I like knowing a type is locked down this way, and I don't have to think about inheritance. Sealing classes can impact perf positively, though more in the modern versions of .NET than in .NET Framework. And I don't think any of the usage of this `Entry` type would benefit from the perf side. Perhaps for the arrays within the dictionary, but that wasn't on my mind when I sealed it. We can revert if you prefer as I don't feel that strongly.

This blog post seems quite good on the perf implications of sealing classes: https://www.meziantou.net/performance-benefits-of-sealed-class.htm

Reply:
Thanks for the link. I'm surprised how much of a difference (in percentage terms) casting can be, though I highly value composition over inheritance, which minimises the frequency that casting is used. I also don't understand why production code would use interfaces or inheritance if it then just casts to a (sealed) concrete type, which makes it feel more academic than practical to me. But it does make nice benchmark results.